src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,459 @@
+            INamedTypeSymbol containingType, ITypeSymbol collectionType, 
+            ITypeSymbol iterationType, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var semanticFacts = document.GetLanguageService<ISemanticFactsService>();
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var editor = new SyntaxEditor(root, generator);
+
+            // create a dummy "list[i]" expression.  We'll use this to find all places to replace
+            // in the current for statement.
+            var indexExpression = generator.ElementAccessExpression(
+                collectionExpression, generator.IdentifierName(iterationVariable));
+
+            // See if the first statement in the for loop is of the form:
+            //      var x = list[i]   or
+            //
+            // If so, we'll use those as the iteration variables for the new foreach statement.
+            var bodyStatements = GetBodyStatements(forStatement);
+            var (typeNode, foreachIdentifier, declarationStatement) = tryDeconstructInitialDeclaration();
+
+            if (typeNode == null)
+            {
+                // user didn't provide an explicit type.  Check if the index-type of hte collection
+                // is different from than .Current type of the enumerator.  If so, add an explicit
+                // type so that the foreach will coerce the types accordingly.
+                var indexerType = GetIndexerType(containingType, collectionType);
+                if (!Equals(indexerType, iterationType))
+                {
+                    typeNode = (TTypeNode)generator.TypeExpression(
+                        indexerType ?? semanticModel.Compilation.GetSpecialType(SpecialType.System_Object));
+                }
+            }
+
+            // If we couldn't find an appropriate existing variable to use as the foreach
+            // variable, then generate one automatically.
+            if (foreachIdentifier.RawKind == 0)
+            {
+                foreachIdentifier = semanticFacts.GenerateUniqueName(
+                    semanticModel, forStatement, containerOpt: null, baseName: "v", cancellationToken);
+                foreachIdentifier = foreachIdentifier.WithAdditionalAnnotations(RenameAnnotation.Create());
+            }
+
+            // Create the expression we'll use to replace all matches in the for-body.
+            var foreachIdentifierReference = foreachIdentifier.WithoutAnnotations(RenameAnnotation.Kind).WithoutTrivia();
+
+            // Walk the for statement, replacing any matches we find.
+            findAndReplaceMatches(forStatement);
```

Comment:
Consider passing `foreachIdentifierReference` into `findAndReplaceMatches` for readability (easier to see what flows in/out).

Reply:
Tried both ways.  i like the current form.